tests/server.rs

```diff
@@ -1482,6 +1482,72 @@ async fn http_connect_new() {
     assert_eq!(s(&vec), "bar=foo");
 }
 
+#[tokio::test]
+async fn h2_connect() {
+    use tokio::io::{AsyncReadExt, AsyncWriteExt};
+    let _ = pretty_env_logger::try_init();
+    let listener = tcp_bind(&"127.0.0.1:0".parse().unwrap()).unwrap();
+    let addr = listener.local_addr().unwrap();
+    let (done_tx, mut done_rx) = tokio::sync::mpsc::channel(10);
+
+    let conn = connect_async(addr).await;
+    let (h2, connection) = h2::client::handshake(conn).await.unwrap();
+    tokio::spawn(async move {
+        connection.await.unwrap();
+    });
+
+    tokio::spawn(async move {
+        let mut h2 = h2.ready().await.unwrap();
+
+        let request = Request::connect("localhost").body(()).unwrap();
+        let (response, mut send_stream) = h2.send_request(request, false).unwrap();
+        let mut response = response.await.unwrap();
+        assert_eq!(response.status(), StatusCode::OK);
+
+        let bytes = response.body_mut().data().await.unwrap().unwrap();
+        assert_eq!(&bytes[..], b"Bread?");
+        let _ = response
+            .body_mut()
+            .flow_control()
+            .release_capacity(bytes.len());
+
+        send_stream.send_data("Baguette!".into(), true).unwrap();
```

Comment:
From previous experience it's worth testing quite long bodies that don't fit in MTU as a whole to ensure that pending reads/write are waken up correctly. And 0 length bodies as a special case.

Also, for the multiplexing case: if one of the streams error does it terminate the rest of the tunnels along with the connection?

Reply:
I added a test with a very large body. I'm not sure what you want me to test for 0 length bodies, as receiving 0 bytes usually signal EOF?

Reply:
I added a test that checks that 0-length non-EOS data frames don't make the AsyncRead impl bail out.